Enable llhttp for HTTP parsing

sgallagher opened the pull request:
Fixes: https://github.com/libgit2/libgit2/issues/6074

yselkowitz commented on src/libgit2/transports/httpclient.c:
```diff
@@ -7,7 +7,36 @@
 #include "git2.h"
-#include "http_parser.h"
+
+#define USE_LLHTTP 1
```
Leftover from devel testing?  This is also probably the cause of the CI failures, although CI seems to use the bundled dependencies so it may not really test how this builds with llhttp.

sgallagher replied:
Yeah, that was a glaring mistake.

sergio-correia commented:
`src/util/net.c` also has an `http_parser.h` include that should likely be removed.

sergio-correia commented on ci/docker/fedora:
```diff
@@ -0,0 +1,61 @@
+	krb5-workstation \
+	krb5-libs \
+    krb5-devel \
+    pcre2-devl \
```
typo: `pcre2-devel`

sergio-correia commented:
@ethomson: Hi, would you mind running the CI for this PR, please, so we see what is failing? Thanks in avance.

sgallagher commented:
Thanks to a lot of help from @sergio-correia, I think this is in proper shape for review. @ethomson if you could kick off the pipeline check, I'd appreciate it.

sgallagher commented:
I can't reproduce that valgrind error locally. Given where it's located in the code, I suspect it may have been a bug in glibc getaddrinf() that has been fixed in the meantime. Could you retry the check?

ethomson commented:
Hiya - many apologies for the delay. Thank you so much for all this — one question I have is whether we really need to support _both_ llhttp and the old Node http_parser? It would simplify the code quite a bit to only support the newfangled llhttp, which we could bundle as a dependency (removing the old Node http_parser).

If there's a reason to keep the old support, I'm willing. Just curious.

sgallagher commented:
> Hiya - many apologies for the delay. Thank you so much for all this — one question I have is whether we really need to support _both_ llhttp and the old Node http_parser? It would simplify the code quite a bit to only support the newfangled llhttp, which we could bundle as a dependency (removing the old Node http_parser).
> 
> If there's a reason to keep the old support, I'm willing. Just curious.

I mostly left it alone because I wasn't prepared to try to entirely extract http-parser since you were carrying a bundled copy. If you're willing to make the transition directly, I'm fully on-board. Upstream http-parser has been dead for four years and I have a sneaking suspicion that there are probably a number of unaddressed security vulnerabilities there that have gone unreported.

ethomson commented:
Thanks for tackling this PR - it's been on my wish-list for a while to move to llhttp for a plurality of reasons.

After thinking about it a bit more, I realized that other distros should be able to continue to use their existing http-parsers. So I kept the option to use http-parser _or_ llhttp. But I added llhttp as the bundled option, and removed the http-parser that we were bundling. I also brought some of our abstractions a bit more in line with libgit2's typical code style. This lets the http client talk to a generic http parser without having to know what kind of http parser it is, and encapsulates the http parser switching logic in http parser code.

Thanks again! Excited to see this.

sgallagher commented:
> Thanks for tackling this PR - it's been on my wish-list for a while to move to llhttp for a plurality of reasons.

Happy to help :)

> After thinking about it a bit more, I realized that other distros should be able to continue to use their existing http-parsers. So I kept the option to use http-parser _or_ llhttp. But I added llhttp as the bundled option, and removed the http-parser that we were bundling. I also brought some of our abstractions a bit more in line with libgit2's typical code style. This lets the http client talk to a generic http parser without having to know what kind of http parser it is, and encapsulates the http parser switching logic in http parser code.

Thanks for get it over the finish line. Obviously it's my first contribution to libgit2, so I appreciate the help.

> Thanks again! Excited to see this.

Us too! This is (I think) the last major package in Fedora that's pulling in http-parser, so getting rid of that dependency here is a big deal for us.

ethomson commented:
> Thanks for get it over the finish line. Obviously it's my first contribution to libgit2, so I appreciate the help.

Appreciate your kind words - I tend to just jump in and meddle in PRs which is not necessarily the most constructive (or polite) way to work.

Can you triple check me and make sure that I didn't break anything for your use case? I'll hit the shiny green button if it's all good.

sgallagher commented:
Everything looks good to me, though I am curious why you moved the CI run for Fedora to Nightly rather than as part of the merge request workflow. Is it just too slow?

ethomson commented:
Ah right. Yeah, I want to keep the CI runs minimal enough to just cover the various things that we support, but still be fast. And then nightlies can be more exhaustive.

There's already more than I would like in the CIs, they're awfully slow. I'm happy to rethink what we have where.
